gcloud-java-compute/src/test/java/com/google/gcloud/compute/ForwardingRuleIdTest.java

```diff
@@ -0,0 +1,130 @@
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.google.gcloud.compute;
+
+import static org.junit.Assert.assertEquals;
+import static org.junit.Assert.assertFalse;
+import static org.junit.Assert.assertNull;
+import static org.junit.Assert.assertSame;
+import static org.junit.Assert.assertTrue;
+
+import org.junit.Rule;
+import org.junit.Test;
+import org.junit.rules.ExpectedException;
+
+public class ForwardingRuleIdTest {
+
+  private static final String PROJECT = "project";
+  private static final String REGION = "region";
+  private static final String NAME = "rule";
+  private static final String GLOBAL_URL =
+      "https://www.googleapis.com/compute/v1/projects/project/global/forwardingRules/rule";
+  private static final String REGION_URL = "https://www.googleapis.com/compute/v1/projects/"
+      + "project/regions/region/forwardingRules/rule";
+
+  @Rule
+  public ExpectedException thrown = ExpectedException.none();
+
+  @Test
+  public void testOf() {
+    GlobalForwardingRuleId forwardingRuleId = GlobalForwardingRuleId.of(PROJECT, NAME);
+    assertEquals(PROJECT, forwardingRuleId.project());
+    assertEquals(NAME, forwardingRuleId.rule());
+    assertEquals(GLOBAL_URL, forwardingRuleId.selfLink());
+    assertEquals(ForwardingRuleId.Type.GLOBAL, forwardingRuleId.type());
+    forwardingRuleId = GlobalForwardingRuleId.of(NAME);
+    assertNull(forwardingRuleId.project());
+    assertEquals(NAME, forwardingRuleId.rule());
+    assertEquals(ForwardingRuleId.Type.GLOBAL, forwardingRuleId.type());
+    RegionForwardingRuleId regionForwardingRuleId =
+        RegionForwardingRuleId.of(PROJECT, REGION, NAME);
+    assertEquals(PROJECT, regionForwardingRuleId.project());
+    assertEquals(REGION, regionForwardingRuleId.region());
+    assertEquals(NAME, regionForwardingRuleId.rule());
+    assertEquals(REGION_URL, regionForwardingRuleId.selfLink());
+    assertEquals(ForwardingRuleId.Type.REGION, regionForwardingRuleId.type());
+    regionForwardingRuleId = RegionForwardingRuleId.of(RegionId.of(PROJECT, REGION), NAME);
+    assertEquals(PROJECT, regionForwardingRuleId.project());
+    assertEquals(REGION, regionForwardingRuleId.region());
+    assertEquals(NAME, regionForwardingRuleId.rule());
+    assertEquals(REGION_URL, regionForwardingRuleId.selfLink());
+    assertEquals(ForwardingRuleId.Type.REGION, regionForwardingRuleId.type());
+    regionForwardingRuleId = RegionForwardingRuleId.of(REGION, NAME);
+    assertNull(regionForwardingRuleId.project());
+    assertEquals(REGION, regionForwardingRuleId.region());
+    assertEquals(NAME, regionForwardingRuleId.rule());
+    assertEquals(ForwardingRuleId.Type.REGION, regionForwardingRuleId.type());
+  }
+
+  @Test
+  public void testToAndFromUrl() {
+    GlobalForwardingRuleId forwardingRuleId = GlobalForwardingRuleId.of(PROJECT, NAME);
```

Comment:
Related to urls/selfLinks: Do you think we should throw some sort of exception if the user asks for a URL/selfLink if they haven't set a project ID?  Or do you think it's better as is?  (I don't feel strongly either way, but I thought I'd ask anyways).